src/sync/mpsc.rs

```diff
@@ -0,0 +1,134 @@
+//! A multi-producer, single-consumer, futures-aware channel
+
+use {Async, Poll};
+use std::sync::atomic::Ordering::SeqCst;
+use std::sync::atomic::AtomicUsize;
+use stream::Stream;
+use task::{self, Task};
+use std::sync::{Arc, mpsc};
+use lock::Lock;
+
+/// `Sender` of the channel. `Sender` can be cloned.
+pub struct Sender<T> {
+    /// Sender
+    tx: mpsc::Sender<T>,
```

Comment:
Eventually we'll probably want to replace this with our own queue (as most of the machinery in std we're not using as we're only using `try_recv` here, but we can always optimize that later.